src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs

```diff
@@ -171,38 +171,49 @@ public async Task<IActionResult> GetCurrentForecastData([FromQuery(Name = "locat
     /// current forecast at the time.
     /// </remarks>
     /// <param name="requestedForecasts"> Array of requested forecasts.</param>
     /// <returns>An array of forecasts with their optimal marginal carbon intensity window.</returns>
     /// <response code="200">Returns the requested forecast objects</response>
-    /// <response code="400">Returned if any of the requested items are invalid</response>
+    /// <response code="400">Returned if any of the input parameters are invalid</response>
     /// <response code="500">Internal server error</response>
     /// <response code="501">Returned if the underlying data source does not support forecasting</response>
     [Produces("application/json")]
-    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(IEnumerable<EmissionsForecastDTO>))]
+    [ProducesResponseType(StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ValidationProblemDetails))]
     [ProducesResponseType(StatusCodes.Status500InternalServerError, Type = typeof(ValidationProblemDetails))]
-    [ProducesResponseType(StatusCodes.Status501NotImplemented, Type = typeof(ValidationProblemDetails))]
     [HttpPost("forecasts/batch")]
-    public IActionResult BatchForecastData(IEnumerable<EmissionsForecastBatchDTO> requestedForecasts)
+    public async IAsyncEnumerable<EmissionsForecastDTO> BatchForecastDataAsync(IEnumerable<EmissionsForecastBatchDTO> requestedForecasts)
     {
-        // Dummy result.
-        // TODO: implement this controller method after spec is approved.
-        var result = new List<EmissionsForecastDTO>();
-        return Ok(result);
+        using (var activity = Activity.StartActivity())
+        {
+            foreach (var forecastBatchDTO in requestedForecasts)
+            {
+                IEnumerable<Location> locationEnumerable = CreateLocationsFromQueryString(new string[] { forecastBatchDTO.Location! });
+                var props = new Dictionary<string, object?>() {
+                    { CarbonAwareConstants.Locations, locationEnumerable },
+                    { CarbonAwareConstants.Start, forecastBatchDTO.DataStartAt },
+                    { CarbonAwareConstants.End, forecastBatchDTO.DataEndAt },
+                    { CarbonAwareConstants.Duration, forecastBatchDTO.WindowSize },
+                    { CarbonAwareConstants.ForecastRequestedAt, forecastBatchDTO.RequestedAt },
+                };
+                // NOTE: Current Error Handling done by HttpResponseExceptionFilter can't handle exceptions
+                // thrown by the underline framework for this method, therefore all exceptions are handled as 500.
+                // Refactoring with a middleware exception handler should cover this use case too.
+                var forecast = await _aggregator.GetForecastDataAsync(props);
+                yield return EmissionsForecastDTO.FromEmissionsForecast(forecast);
```

Comment:
the response object now expects the `forecastBatchDTO.RequestedAt` to be included alongside the other `DataStartAt`, `DataEndAt`, `GeneratedAt`, etc..., for each item in the batch

Reply:
yes, that is done by the static method `FromEmissionsForecast`

Reply:
hum, are you sure @juzuluag -- I [checked here](https://github.com/microsoft/carbon-aware-sdk/blob/390/hist_forecast_impl/src/CarbonAware.WebApi/src/Models/EmissionsForecastDTO.cs), and didn't see it doing so, and my guess is the internal `EmissionsForecast` object passed into `FromEmissionsForecast` wasn't updated to include `requestedAt` ?

Reply:
Good catch @dnastrain . Fixed with https://github.com/microsoft/carbon-aware-sdk/pull/110/commits/d793ab53d57b7936578e43fa8d7e0a7b37ec0d9a

Reply:
I added comment further down @juzuluag , since while I agree the change above does "fix it", I think there's a way that follows the existing code-pattern more closely ...